File: netlogo-gui/src/main/prim/_turtlevariabledouble.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim;

import org.nlogo.agent.Turtle;
import org.nlogo.api.AgentException;
import org.nlogo.api.LogoException;
import org.nlogo.core.Syntax;
import org.nlogo.nvm.Context;
import org.nlogo.nvm.RuntimePrimitiveException;
import org.nlogo.nvm.Reporter;

public final class _turtlevariabledouble extends Reporter {
  public int vn;



  @Override
  public String toString() {
    if (world == null) {
      return super.toString() + ":" + vn;
    }
    return super.toString() + ":" + world.turtlesOwnNameAt(vn);
  }

  @Override
  public Object report(Context context) throws LogoException {
    try {
      return context.agent.getTurtleVariable(vn);
    } catch (AgentException ex) {
      throw new RuntimePrimitiveException(context, this, ex.getMessage());
    }
  }

  public double report_1(Context context) {
    return ((Turtle) context.agent).getTurtleVariableDouble(vn);
  }

  public Double report_2(Context context) throws LogoException {
    try {
      return (Double) context.agent.getTurtleVariable(vn);
    } catch (AgentException ex) {
      throw new RuntimePrimitiveException(context, this, ex.getMessage());
    }
  }
}

File: netlogo-headless/src/main/prim/_turtlevariabledouble.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim

import org.nlogo.agent.Turtle
import org.nlogo.api.AgentException
import org.nlogo.nvm.{ Context, Reporter }
import org.nlogo.nvm.RuntimePrimitiveException

class _turtlevariabledouble(private var _vn: Int) extends Reporter {

  def this() = this(0)

  override def toString =
    super.toString + ":" +
      Option(world).map(_.turtlesOwnNameAt(vn)).getOrElse(vn.toString)

  // MethodRipper won't let us call a public method from report_1()
  // so we must keep vn and _vn separate - ST 9/22/12
  def vn = _vn
  def vn_=(vn: Int): Unit = { _vn = vn }

  override def report(context: Context): java.lang.Double =
    report_2(context)

  def report_1(context: Context): Double =
    context.agent.asInstanceOf[Turtle].getTurtleVariableDouble(_vn)

  def report_2(context: Context): java.lang.Double =
    try context.agent.getTurtleVariable(_vn).asInstanceOf[java.lang.Double]
    catch { case ex: AgentException =>
      throw new RuntimePrimitiveException(context, this, ex.getMessage) }

}
